### PR TITLE
u3: use new road-stack api in noun traversals

### PR DESCRIPTION
**Pull request by joemfb**
Building on #3433, this PR refactors/rewrites our big, open-ended noun traversals (pre-order in `u3a_walk_fore/unsafe()` and `u3r_sing()`, post-order in `u3r_mug()`, `u3s_cue()`, `u3a_take()`).

Their manual stack fiddling (and road-direction abstraction) have been replaced with the new road-stack API, and the post-order traversals use two functions with (do) while loops instead of labeled goto. I've also generally cleaned up and commented the implementations, and split `u3a_take()` into north/south variants (it used to work this way, but the old explicit-stack version was so big that I couldn't bring myself to duplicate it).

### CODE REVIEW

**Comment by joemfb** [comment]
As of this PR, all usage of the road stack goes through the new, except for the bytecode interpreter, which has special requirements and uses many more operations.

File: pkg/urbit/noun/retrieve.c [hunk 1]
```diff
@@ -221,14 +221,14 @@ u3r_mean(u3_noun som, ...)
 
 //  stack frame for tracking noun comparison and unification
 //
-//    We're always comparing arbitrary nouns in a %none frame.
-//    When we comparing two cells, we change the none frame to
-//    a head-frame and push a new %none frame for the heads.
-//    If the heads were equal, the head-frame has the context to
-//    unify their head pointers. We then repeat with the tails,
+//    we always comparing arbitrary nouns in a none-frame.
```
Comment (eglaysher) [inline pkg/urbit/noun/retrieve.c:224]: tense

**Review by eglaysher — APPROVED** [review]
My eyes have rolled over every line of this patch, for what that's worth, though this patch is so large and finicky, this approval should be treated as a rubber stamp.

File: pkg/urbit/noun/allocate.c [hunk 2]
```diff
@@ -1237,19 +1239,23 @@ u3a_take(u3_noun veb)
...
 
         fam_u->tag_y = TAKE_TAIL;
         fam_u->hed   = pro;
 
-        veb = old_u->tel;
-        goto advance;
+        // veb = old_u->tel;
+        // goto advance;
+        pro = _ca_take_next(&pil_u, old_u->tel);
+        goto retreat;
```
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/urbit/noun/allocate.c]: The `advance` block became `_ca_take_next()`, so this is just a manual loop `continue`, previously implicit in the fact that the `retreat` block was after the `advance` block.

[the remark this answers is not in the document]
Comment (eglaysher) [reply pkg/urbit/noun/allocate.c]: ok

File: pkg/urbit/noun/allocate.c [hunk 3]
```diff
@@ -1064,6 +1064,36 @@ _ca_take_atom(u3a_atom* old_u)
   return new;
 }
 
+/* _ca_take_cell(): reallocate a cell atom off the stack.
```
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/urbit/noun/allocate.c]: fixed

File: pkg/urbit/noun/retrieve.c (hunk 1 above)
Comment (joemfb) [reply pkg/urbit/noun/retrieve.c]: fixed

**Comment by joemfb** [comment]
Thanks, @eglaysher, understood. These operations are fundamental and used incredibly frequently, so passing the CI test (under U3_MEMORY_DEBUG, in the pr build) is a pretty strong indicator of correctness.

**Pull request merged**